### DISCUSSION TITLE
Redirect issue with getserversideprops

### DISCUSSION THREAD

**Discussion by horlaarsco** [vercel/next.js#13871] [Help]
I'm trying to redirect in login page to the home page when the cookie that verifies authentication is found.  when I use getserversideprops... Page redirects on first load but when i hit refresh of the login page again i get error of header sent.. screenshot below.

![image](https://user-images.githubusercontent.com/10059733/83976802-3309fa80-a8f4-11ea-9426-97337b786c5b.png)

![image](https://user-images.githubusercontent.com/10059733/83977132-00f99800-a8f6-11ea-8698-aa49dafc5652.png)

**Comment by marko-hologram** [comment]
From your question, I'm not entirely sure what you mean, but the first problem I notice in your screenshot is that you are using `getServerSideProps` inside a component file, which is something you cannot do.

https://nextjs.org/docs/basic-features/data-fetching#only-allowed-in-a-page-2

If you take a look in the docs here, you will notice it says it can only be exported from a page file, so that's the first thing I would update in your example.

**Reply by horlaarsco** [reply]
please check now i uploaded the wrong image..

**Reply by marko-hologram** [reply]
Hmm to be honest I'm not sure why it's failing. I just tested it locally with a small code example and it works. Are you maybe handling some server side props in the global `_app` or `_document` which might clash with the logic you have here?

This is the small example I used, basically the same as your code.

```jsx
import cookie from 'cookie';
import Head from 'next/head';
import Link from 'next/link';
import { useEffect } from 'react';

function setCookie(cname, cvalue, exdays) {
  var d = new Date();
  d.setTime(d.getTime() + exdays * 24 * 60 * 60 * 1000);
  var expires = 'expires=' + d.toUTCString();
  document.cookie = cname + '=' + cvalue + ';' + expires + ';path=/';
}

export default function Login({ name = '' }) {
  useEffect(() => {
    setCookie('TEST_COOKIE', true);
  }, []);

  return (
    <div className="container">
      <Head>
        <title>Login Page</title>
      </Head>

      <h1>Hello {name}</h1>
      <Link href="/" as="/">
        <a>Index</a>
      </Link>
    </div>
  );
}

export const getServerSideProps = async ({ req, res }) => {
  const cookies = cookie.parse(req.headers.cookie ?? '');
  if (cookies['TEST_COOKIE']) {
    res.writeHead(302, { Location: '/' });
    res.end();
    return { props: {} };
  }

  return { props: { name: 'Test' } };
};

```

**Reply by horlaarsco** [reply]
I'm handling some in the _app.js with getInitialProps.. As getserversideprops doesn't work there

**Reply by marko-hologram** [reply]
Are you maybe doing `res.end()` there or something else? 

Because the error message "...write after end" means that you tried to write something to the response after the response has ended, so your `res.writeHead()` is executed too late.

**Reply by horlaarsco** [reply]
![image](https://user-images.githubusercontent.com/10059733/83978037-a44dab80-a8fc-11ea-9495-e9c667ddf123.png)

**Reply by Felipithstdr** [reply]
Where do I put this code you made?

**Comment by timneutkens** [comment]
Opened a RFC here: https://github.com/vercel/next.js/discussions/14890
